File: src/content.ts
import * as fs from "node:fs/promises";
import * as path from "node:path";
import log from "loglevel";
import { parse } from "smol-toml";
import { z } from "zod";
import { PATHS } from "./constants.js";
import { render_djot } from "./djot.js";
import { minifyHTML } from "./minify.js";
import { Path } from "./path.js";
import { Site } from "./site.js";
import { render } from "./templating.js";
import { ensureDir } from "./utils.js";
import type { WriteFromSite } from "./writeable.js";

const FRONTMATTER_DELIMITER = "+++";

export const Frontmatter = z
	.object({
		title: z.string(),
		last_modified: z.date(),
		description: z.string(),
		subtitle: z.string().optional(),
		slug: z.string().optional(),
		layout: z.enum(["page", "index", "404"]).default("page"),
		special: z.boolean().default(false),
		draft: z.boolean().default(false),
	})
	.transform(({ last_modified, ...rest }) => ({ ...rest, lastModified: last_modified }));

export type Frontmatter = z.infer<typeof Frontmatter>;

export class Content implements WriteFromSite {
	public sourcePath: Path;
	public url: URL;
	public contentType: "page" | "post";
	public frontmatter: Frontmatter;

	private sourceContent: string;

	private constructor(source: Path, url: URL, contentType: "page" | "post", frontmatter: Frontmatter, body: string) {
		this.sourcePath = source;
		this.url = url;
		this.contentType = contentType;
		this.frontmatter = frontmatter;
		this.sourceContent = body;
	}

	public get outPath(): Path {
		if (this.frontmatter.slug === undefined) {
			return new Path(path.join(PATHS.out, this.sourcePath.stem, "index.html"));
		}

		return new Path(path.join(PATHS.out, this.frontmatter.slug, "index.html"));
	}

	public get content() {
		return render_djot(this.sourceContent);
	}

	public context(site: Site) {
		return {
			title: `${this.frontmatter.title} => Eons :: IO ()`,
			canonicalUrl: new URL(this.url.toString(), site.url).toString(),
			css: site.assets.get("styles.css"),
			isDev: !site.isProd,
			content: this.content,
			frontmatter: this.frontmatter,
			pubDate: this.frontmatter.lastModified?.toISOString(),
		};
	}

	public async write(site: Site) {
		await ensureDir(path.dirname(this.outPath.absolute));
		let rendered = await render(this.frontmatter.layout, this.context(site));

		if (site.isProd) {
			rendered = await minifyHTML(rendered);
		}

		await fs.writeFile(this.outPath.absolute, rendered);
	}

	public static async fromPath(filePath: string, kind: "page" | "post", baseURL: URL): Promise<Content> {
		const source = await fs.readFile(filePath, "utf-8");

		const start = source.indexOf(FRONTMATTER_DELIMITER);
		const end = source.indexOf(FRONTMATTER_DELIMITER, start + FRONTMATTER_DELIMITER.length);
		const attrs = source.substring(start + FRONTMATTER_DELIMITER.length, end).trim();
		const body = source.substring(end + FRONTMATTER_DELIMITER.length).trim();

		const toml = parse(attrs);
		const frontmatter = Frontmatter.safeParse(toml);

		if (!frontmatter.success) {
			log.error(frontmatter.error);
			throw new Error(`Failed to parse frontmatter for ${filePath}`);
		}

		const stem = `${path.parse(filePath).name}/`;
		const url = new URL(frontmatter.data.slug ?? stem, baseURL);

		return new Content(new Path(filePath), url, kind, frontmatter.data, body);
	}
}

File: src/content.rs
use std::path::{Path, PathBuf};

use anyhow::{Context, Result};
use jotdown::{Attributes, Container, Event, Render};
use minijinja::{context, value::Value};
use minijinja_autoreload::AutoReloader;
use serde::Deserialize;
use syntect::{easy::HighlightLines, highlighting::ThemeSet, parsing::SyntaxSet};
use time::Date;
use url::Url;

use crate::{utils::toml_date_deserializer, Mode};

#[derive(Debug, Deserialize)]
pub struct Frontmatter {
    pub title: String,
    #[serde(with = "toml_date_deserializer")]
    pub last_modified: Date,
    pub subtitle: Option<String>,
    pub description: String,
    pub slug: Option<String>,
    pub layout: Option<String>,
    #[serde(default)]
    pub special: bool,
}

#[derive(Debug, Copy, Clone)]
pub enum Type {
    Page,
    Post,
}

#[derive(Debug)]
pub struct Content {
    pub source: PathBuf,
    pub out_path: PathBuf,
    pub url: String,
    pub content_type: Type,
    pub frontmatter: Frontmatter,
    pub content: String,
}

impl Content {
    pub fn from_path(path: &Path, kind: Type) -> Result<Self> {
        let file = std::fs::read_to_string(path)?;
        let stem = path.file_stem().unwrap().to_string_lossy();

        match file
            .split("+++")
            .map(str::trim)
            .filter(|e| !e.is_empty())
            .collect::<Vec<_>>()[..]
        {
            [frontmatter, content] => {
                Content::from_file(path, &stem, kind, frontmatter, Some(content))
            }
            [frontmatter] => Content::from_file(path, &stem, kind, frontmatter, None),
            _ => todo!(),
        }
    }

    pub fn render(
        &self,
        styles: &str,
        mode: Mode,
        url: &Url,
        templates: &AutoReloader,
    ) -> Result<String> {
        let env = templates.acquire_env()?;
        let template = env.get_template(&self.layout())?;
        let context = self.create(styles, mode, url)?;
        template
            .render(context)
            .context("Failed to render template")
    }

    fn from_file(
        source: &Path,
        stem: &str,
        kind: Type,
        frontmatter: &str,
        content: Option<&str>,
    ) -> Result<Self> {
        let frontmatter: Frontmatter = toml::from_str(frontmatter)?;

        let path: PathBuf = match &frontmatter.slug {
            Some(slug) => [slug, "index.html"].into_iter().collect(),
            None => [stem, "index.html"].into_iter().collect(),
        };

        let url = frontmatter.slug.as_ref().map_or(stem, |s| s);

        Ok(Content {
            source: source.to_path_buf(),
            url: format!("{url}/"),
            out_path: path,
            content_type: kind,
            content: content.unwrap_or_default().into(),
            frontmatter,
        })
    }

    pub fn filename(&self) -> String {
        self.source.file_stem().map_or_else(
            || panic!("No filename found"),
            |name| name.to_string_lossy().to_string(),
        )
    }

    fn layout(&self) -> String {
        match (self.content_type, &self.frontmatter.layout) {
            (_, Some(layout)) => format!("{layout}.jinja"),
            (Type::Page, None) => "page.jinja".to_string(),
            (Type::Post, None) => "post.jinja".to_string(),
        }
    }

    fn content(&self) -> Result<String> {
        let events = jotdown::Parser::new(&self.content).map(jotdown_event_mapper);
        let mut html = String::new();
        jotdown::html::Renderer::default().push(events, &mut html)?;
        Ok(html)
    }

    fn create(&self, styles: &str, mode: Mode, url: &Url) -> Result<Value> {
        let content = self.content()?;

        Ok(context! {
            title => self.frontmatter.title.clone(),
            subtitle => self.frontmatter.subtitle.clone(),
            description => self.frontmatter.description.clone(),
            is_dev => mode.is_dev(),
            canonical_url => url.join(&self.url)?,
            content => content,
            styles => styles,
        })
    }
}

fn jotdown_event_mapper(event: Event) -> Event {
    match event {
        Event::Start(container, attrs) => jotdown_container_mapper(container, attrs).into(),
        _ => event,
    }
}

struct ContainerWrapper<'a> {
    container: Container<'a>,
    attrs: Attributes<'a>,
}

impl<'a> From<ContainerWrapper<'a>> for Event<'a> {
    fn from(val: ContainerWrapper<'a>) -> Self {
        Event::Start(val.container, val.attrs)
    }
}

fn jotdown_container_mapper<'a>(
    container: Container<'a>,
    attrs: Attributes<'a>,
) -> ContainerWrapper<'a> {
    let ps = SyntaxSet::load_defaults_newlines();
    let ts = ThemeSet::load_defaults();

    match container {
        Container::Heading {
            id,
            level,
            has_section,
        } => ContainerWrapper {
            container: Container::Heading {
                level,
                id: id.to_lowercase().into(),
                has_section,
            },
            attrs,
        },
        Container::Section { id } => ContainerWrapper {
            container: Container::Section {
                id: id.to_lowercase().into(),
            },
            attrs,
        },
        Container::CodeBlock { language } => {
            let syntax = ps.find_syntax_by_name("JavaScript").unwrap();
            let theme = &ts.themes["InspiredGitHub"];
            ContainerWrapper {
                container: Container::CodeBlock { language },
                attrs,
            }
        }
        _ => ContainerWrapper { container, attrs },
    }
}
